lib/internal/Magento/Framework/View/Page/Config/Renderer.php

```diff
@@ -344,43 +344,73 @@ protected function getGroupAttributes($group)
      * Add default attributes
      *
      * @param string $contentType
-     * @param string $attributes
-     * @return string
+     * @param array|null $attributes
+     * @return array
      */
     protected function addDefaultAttributes($contentType, $attributes)
     {
+        $attributesString = '';
+        $defaultAttributes = [];
+
         if ($contentType === 'js') {
-            return ' type="text/javascript" ' . $attributes;
+            $defaultAttributes['type'] = 'text/javascript';
         }
 
         if ($contentType === 'css') {
-            return ' rel="stylesheet" type="text/css" ' . ($attributes ?: ' media="all"');
+            $defaultAttributes['rel'] = 'stylesheet';
+            $defaultAttributes['type'] = 'text/css';
+
+            if (empty($attributes)) {
+                $attributes['media'] = 'all';
+            }
+
+            // add defer attributes when defer parameter is exist
+            if (is_array($attributes) && array_key_exists('defer', $attributes)) {
+                $defaultAttributes['rel'] = 'preload';
+                $defaultAttributes['as'] = 'style';
+                $defaultAttributes['onload'] = 'this.onload=null;this.rel=\'stylesheet\'';
+            }
         }
 
         if ($this->canTypeBeFont($contentType)) {
-            return 'rel="preload" as="font" crossorigin="anonymous"';
+            $defaultAttributes['rel'] = 'preload';
+            $defaultAttributes['as'] = 'font';
+            $defaultAttributes['crossorigin'] = 'anonymous';
         }
 
-        return $attributes;
+        // merge current attributes with default attributes
+        if ($attributes) {
+            return array_merge($defaultAttributes, $attributes);
+        }
+
+        return $defaultAttributes;
     }
 
     /**
      * Returns assets template
      *
      * @param string $contentType
-     * @param string|null $attributes
+     * @param array|null $attributes
      * @return string
      */
     protected function getAssetTemplate($contentType, $attributes)
     {
+        $attributesString = '';
+        foreach ($attributes as $name => $value) {
+            $attributesString .= ' ' . $name . '="' . $this->escaper->escapeHtml($value) . '"';
```

Comment:
Should be `escapeHtmlAttr` instead of `escapeHtml` I think?

----

And in the case of the `onload` one, it should be `escapeJs` inside a `escapeHtmlAttr` according to [the docs](https://developer.adobe.com/commerce/php/development/security/cross-site-scripting/):
> Case: All JavaScript inside attributes must be escaped by escapeJs before escapeHtmlAttr:

Reply:
The `escapeHtmlAttr` function escapes slash character and it looks like the below;

![image](https://github.com/magento/magento2/assets/10761479/30f4c19e-76bf-42ab-b6be-91b4b2ac742a)

it's pretty useful for body element attributes but not for head element attributes. Also, before this pull request, attribute value was cleaned with `escapeHtml` function.

https://github.com/magento/magento2/blob/0a834cce01c3d924d450c9a71123d9ac6de68b62/lib/internal/Magento/Framework/View/Page/Config/Renderer.php#L332-L334

Reply:
Thanks for the feedback! I know too little about this, would be great if somebody with a little bit of security knowledge could double check this here then.

Reply:
If the main issue is that escapeHtmlAttr escaped all chars then that is correct behavior even for head. 

For example:
```
<html>
<head>
<style id=&quot;foo&amp;#x2F;bar&quot;></style>
</head>
<body>
Test succeeds if this is green
<script>document.getElementById('foo/bar').innerHTML = 'body { color: green; }';</script>
</body>
</html>
```

The text would be green.

[Here is a small demo](https://cdpn.io/nathanjosiah/fullpage/JjeVyLW?nocache=true&view=fullpage) that shows this off.

Reply:
> If the main issue is that escapeHtmlAttr escaped all chars then that is correct behavior even for head.
> 
> For example:
> 
> ```
> <html>
> <head>
> <style id="foo&#x2F;bar"></style>
> </head>
> <body>
> Test succeeds if this is green
> <script>document.getElementById('foo/bar').innerHTML = 'body { color: green; }';</script>
> </body>
> </html>
> ```
> 
> The text would be green.
> 
> [Here is a small demo](https://cdpn.io/nathanjosiah/fullpage/JjeVyLW?nocache=true&view=fullpage) that shows this off.

Thanks for your review. Yes, it works even if it looks ugly on the code side. But the question is, should we use it? Renderer.php didn't use escapeHtmlAttr before, it was using the escapeHtml function to clear the head attributes value the same as before.

Reply:
All HTML attributes should be using `escapeHtmlAttr` so it looks like this was using the incorrect escaper previously.

Reply:
> All HTML attributes should be using `escapeHtmlAttr` so it looks like this was using the incorrect escaper previously.

When I changed to `escapeHtmlAttr`, I got an empty value in the unit test. I tried on my local environment and GitHub codespace but got the same error. Can you try this?

![image](https://github.com/magento/magento2/assets/10761479/d95076a1-9d04-46f4-8ab3-b24471d97eab)

Reply:
Since this is a unit test and not an integration test the escaper is (appropriately) a mock and not the real escaper. You would need to adjust the escaper mocked methods at the top at line 111 with your code. Looks like this:
```
        $this->escaperMock = $this->getMockBuilder(Escaper::class)
            ->disableOriginalConstructor()
            ->getMock();
        $this->escaperMock->expects($this->any())
            ->method('escapeHtml')
            ->willReturnArgument(0);
```

Reply:
> 

Thanks for your help. I send a commit.

Reply:
Do we need to account for the `onload` attribute that contains javascript with an extra `escapeJs` call as well?

Also, we should best update `escapeHtml` to `escapeHtmlAttr` as well [in the `getGroupAttributes` method](https://github.com/magento/magento2/blob/0fe805f90119dbef44156e7641eb903ce9e837d9/lib/internal/Magento/Framework/View/Page/Config/Renderer.php#L332-L334) like you pointed out earlier.

Reply:
If I wrap it with `escapeJS` function it won't work. I tested it in my local environment, and that CSS file doesn't work on the browser.

I changed the `escapeHtml` in getGroupAttributes value to `escapeHtmlAttr`.

Reply:
Hi @hostep. Is there any update on this PR? I'd appreciate your feedback.

Reply:
I'm not sure how to move forward here, maybe @nathanjosiah still has some thoughts about escaping inline JS which is introduced in this PR?

Reply:
`escapeHtmlAttr` is appropriate for all attributes including script attributes. You would only need to combine `escapeJs` with `escapeHtmlAttr` if there is additional content being injected into the JS. For example [from our devdocs](https://developer.adobe.com/commerce/php/development/security/cross-site-scripting/)

```
<div
    onclick="<?= $escaper->escapeHtmlAttr('handler("' . $escaper->escapeJs($aString) . '", ' . $escaper->escapeJs($aVar) .')') ?>">
    My DIV
</div>
```

As far as I can see, this PR is using the escaper methods correctly.

Reply:
Looks like this PR is still waiting for your review @hostep. If there are no other problems, can you take this to the next step? Thanks.

Reply:
I don't have a lot of expert knowledge on the subject but since somebody forced me to review this (I wasn't able to unassign myself), I approved it anyways.

Reply:
@sky-hub: and how is that a problem? A browser should be able to decode those without an issue. It's like writing `&gt;` or `&lt;` in the html source to display `>` or `<`. In this case `&#x2F` translates to `/` for example. It's redundant sure, but it shouldn't break anything.